runtime/sema/check_assignment.go

```diff
@@ -110,6 +113,95 @@ func (checker *Checker) checkAssignment(
 	return
 }
 
+func (checker *Checker) enforceViewAssignment(assignment ast.Statement, target ast.Expression) {
+	if !checker.CurrentPurityScope().EnforcePurity {
+		return
+	}
+
+	isWriteableInViewContext := func(t Type) bool {
+		// We have to prevent any writes to references, since we cannot know where the value
+		// pointed to by the reference may have come from. Similarly, we can never safely assign
+		// to a resource; because resources are moved instead of copied, we cannot currently
+		// track the origin of a write target when it is a resource. Consider:
+		// -----------------------------------------------------------------------------
+		// pub resource R {
+		//   pub(set) var x: Int
+		//   init(x: Int) {
+		//     self.x = x
+		//   }
+		//  }
+		//
+		// view fun foo(_ f: @R): @R {
+		//   let b <- f
+		//   b.x = 3 // b was created in the current scope but modifies the resource value
+		//   return <-b
+		// }
+		_, isReference := t.(*ReferenceType)
+		return !isReference && !t.IsResourceType()
+	}
+
+	var baseVariable *Variable
+	var accessChain []Type = make([]Type, 0)
+	var inAccessChain = true
+
+	// seek the variable expression (if it exists) at the base of the access chain
+	for inAccessChain {
+		switch targetExp := target.(type) {
+		case *ast.IdentifierExpression:
+			baseVariable = checker.valueActivations.Find(targetExp.Identifier.Identifier)
+			if baseVariable != nil {
+				accessChain = append(accessChain, baseVariable.Type)
+			}
+			inAccessChain = false
+		case *ast.IndexExpression:
+			target = targetExp.TargetExpression
+			elementType := checker.visitIndexExpression(targetExp, true)
```

Comment:
I assume the `targetExp.TargetExpression` is already visited once by the time we reach here? 

If what is needed here is only the type info, maybe look it up from `checker.Elaboration.IndexExpressionTypes[targetExp]` 

Same for member expression as well.

Reply:
Member expression already has a cache check at the top of the function, so if we've already visited the member we should return early with the information we want. Good catch on the target expression though.